### PR TITLE
torchdata/datapipes/iter/load/fsspec: added fsspec datapipes

### LINKED ISSUE #114: fsspec support
## 🚀 Feature

Add a new loader similar to the iopath loader that uses fsspec.

https://github.com/pytorch/data/blob/main/torchdata/datapipes/iter/load/iopath.py

https://filesystem-spec.readthedocs.io/en/latest/

## Motivation

It would be nice to have fsspec in addition to iopath for loading data from general data sources. A lot of projects already use it and support it which makes it a good to add to torchdata as well for uniform support.

PyTorch Lighting, Tensorboard and TorchX have support for fsspec already. It's quite easy to add support for a new storage provider and has many commons ones available already. Internally there's a Manifold provider which is used with many PyTorch/STL projects.

## Alternatives

For common storage providers such as s3 there's generally already support for that in most projects though for custom / less used storage providers a user would have to implement support for each different system. iopath does provide a similar abstraction but it seems like fsspec generally has more OSS adoption so would be nice to have a unified interface across pytorch projects

### PR DESCRIPTION
**Pull request by d4l3k**
This adds fsspec datapipes that are equivalent to the existing iopath ones. The tests are largely equivalent and test the `file://` and `memory://` filesystems to ensure compatibility.

Closes #114

Test plan:

  pytest tests/test_fsspec.py

### CODE REVIEW

**Comment by d4l3k** [comment]
@ejguan can you review this? Thanks!

**Review by ejguan — COMMENTED** [review]
Could you please install `fsspec` for the CI script to turn on the tests?
https://github.com/pytorch/data/blob/6309113c68c1334bd3e352beba0557ddb71f7ce2/.github/workflows/ci.yml#L47

You may need to install `aiohttp` as well.

File: torchdata/datapipes/iter/load/fsspec.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        else:
+            for file_name in fs.ls(path):
+                if not match_masks(file_name, self.masks):
+                    continue
+
+                # ensure the file name has the full fsspec protocol path
+                if file_name.startswith(fs.protocol):
+                    yield file_name
+                else:
+                    abs_path = posixpath.join(path, file_name)
```
[the remark this answers is not in the document]
Comment (d4l3k) [reply torchdata/datapipes/iter/load/fsspec.py:64]: Hopefully we can delete this code once the changes mentioned in https://github.com/fsspec/filesystem_spec/issues/747#issuecomment-989049627 land. 

I can add the os.path for local but I don't have a windows install I can test with.

[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/datapipes/iter/load/fsspec.py:64]: If you can add `fsspec` to the `ci.yml` file as I suggested above, you can rely on CI to test it for windows.

[the remark this answers is not in the document]
Comment (d4l3k) [reply torchdata/datapipes/iter/load/fsspec.py]: got it passing after a few tries

**Review by ejguan — APPROVED** [review]
LGTM, thank you!

File: torchdata/datapipes/iter/load/fsspec.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        filepath_fn: Function that takes in metadata nad returns the target path of the new file
+    """
+
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe[Tuple[Any, U]],
+        mode: str = "w",
+        filepath_fn: Optional[Callable] = None,
+        *,
+        pathmgr=None,
```
Comment (ejguan) [inline torchdata/datapipes/iter/load/fsspec.py:124]: nit: you can remove this part. This is the specific API for `iopath`

File: torchdata/datapipes/iter/load/fsspec.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                if file_name.startswith(fs.protocol):
+                    yield file_name
+                else:
+                    abs_path = posixpath.join(path, file_name)
+                    if self.root.startswith(fs.protocol):
+                        yield fs.protocol + "://" + abs_path
+
+
+@functional_datapipe("load_file_by_fsspec")
+class FSSpecFileLoaderIterDataPipe(IterDataPipe[Tuple[str, StreamWrapper]]):
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/datapipes/iter/load/fsspec.py:78]: Good catch. @d4l3k Could you please also change it to
```suggestion
@functional_datapipe("open_file_by_fsspec")
class FSSpecFileOpenerIterDataPipe(IterDataPipe[Tuple[str, StreamWrapper]]):
```

@NivekT  BTW, the PR to switch existing DataPipe may be required to work on both internal and external codebase.

File: torchdata/datapipes/iter/__init__.py [hunk 4]
```diff
@@ -31,6 +31,11 @@
     IoPathFileLoaderIterDataPipe as IoPathFileLoader,
     IoPathSaverIterDataPipe as IoPathSaver,
 )
+from torchdata.datapipes.iter.load.fsspec import (
+    FSSpecFileListerIterDataPipe as FSSpecFileLister,
+    FSSpecFileLoaderIterDataPipe as FSSpecFileLoader,
+    FSSpecSaverIterDataPipe as FSSpecSaver,
+)
```
Comment (ejguan) [inline torchdata/datapipes/iter/__init__.py:38]: Could you please also add these class to `__all__`?

Comment (d4l3k) [reply torchdata/datapipes/iter/__init__.py:38]: done

File: torchdata/datapipes/iter/load/fsspec.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (d4l3k) [reply torchdata/datapipes/iter/load/fsspec.py]: done

File: torchdata/datapipes/iter/load/fsspec.py (hunk 2 above)
Comment (d4l3k) [reply torchdata/datapipes/iter/load/fsspec.py]: done

**Review by ejguan — APPROVED** [review]
Thanks again!

**Pull request closed without merging**